### PR TITLE
EAHW-1727 Smart Time Entry

### PR DESCRIPTION
**Pull request by LaurieBennett**
Changed Regex to allow for more varied types of times
Changed method of setting ownerId to userId to remove asynchronicity

### CODE REVIEW

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx [hunk 1]
```diff
@@ -386,7 +416,7 @@ describe('EditShiftHours', () => {
     }
   );
 
-  test.each(['00:00', '08:00', '23:59', '04:26', '0000'])(
+  test.each(validTimes)(
```
Comment (daleelR) [inline src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx:419]: What exactly is the ruling on am/pm in the input box?
It isn't mentioned in the ticket, but the test suggests that its permissable. 
However it's not functioning correctly as per earlier comment. 
I think the reason this test managed to pass is because it's checking for the validation message but the service message is the one that's actually being displayed.

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx [hunk 2]
```diff
@@ -26,6 +27,34 @@ const timecardService = require('../../../api/services/timecardService');
...
+  '!',
+  '25',
+  '13am',
+  '24pm',
+  '3p',
+  '6m',
+  '7a',
+];
+
+const validTimes = [
```
Comment (daleelR) [inline src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx:43]: can we add an AM double digit time here. 
e.g. 10

File: src/components/common/validation/time-format/ValidatedTimeEntry.jsx [hunk 3]
```diff
@@ -29,7 +29,7 @@ const ValidatedTimeEntry = ({
           message: errorMessage,
         },
         pattern: {
-          value: /^([01]\d|2[0-3]):?([0-5]\d)$/,
+          value: /^(\d|[0|1]\d|2[0-3])(:?[0-5]\d)?$|^(\d|[1][0-2])(am|pm)$/,
```
[the remark this answers is not in the document]
Comment (LaurieBennett) [reply src/components/common/validation/time-format/ValidatedTimeEntry.jsx]: I've moved away from the regex, but also removed the attempts to handle AM and 3 digit times, it was a little ambitious to attempt to handle them with just one regex and they aren't specified as necessary in the ticket

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx (hunk 1 above)
Comment (LaurieBennett) [reply src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx:408]: As above, I've removed attempting to handle am/pm in the input box, we can add handling for them later

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx (hunk 2 above)
Comment (LaurieBennett) [reply src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx]: As above, I've removed attempting to handle am/pm in the input box

File: src/components/common/validation/time-format/ValidatedTimeEntry.jsx [hunk 4]
```diff
@@ -8,7 +8,22 @@ const ValidatedTimeEntry = ({
   register,
   isRequired,
 }) => {
-  const errorMessage = `You must enter a ${timeType} in the HH:MM 24 hour clock format`;
+  const validateTime = (time) => {
```
Comment (daleelR) [inline src/components/common/validation/time-format/ValidatedTimeEntry.jsx:11]: nit: ever so slightly misleading function name **given** the return value. It suggests the boolean represents if the time was validated or not. 
e.g. isTimeValid / timeValid would be a bit more explicit.
just a suggestion but feel free to just resolve this.

**Review by daleelR — APPROVED** [review]

**Review by holly-blues — COMMENTED** [review]
Just a question - does the backend handle these smart entries already?

File: src/components/common/validation/time-format/ValidatedTimeEntry.jsx [hunk 5]
```diff
@@ -8,7 +8,22 @@ const ValidatedTimeEntry = ({
...
+    if (time.length < 3 && time.length > 0) {
+      const timeRegEx = /^(\d|[01]\d|2[0-3])$/;
```
Comment (holly-blues) [inline src/components/common/validation/time-format/ValidatedTimeEntry.jsx:13]: Could these two regexes be given more descriptive names based on what they do/their purpose? 

Also, if they could be reused somewhere else in the application perhaps lifted out of here and moved into a shared location? 

If moved out it would be good to see some tests just for these regexes. It would help describe their purpose and what they do e.g. for someone fresh to the project/future maintainers of the project

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx [hunk 6]
```diff
@@ -405,11 +424,12 @@ describe('EditShiftHours', () => {
...
-        const errorMessage = screen.queryByText(
-          'You must enter a finish time in the HH:MM 24 hour clock format'
-        );
-        expect(errorMessage).toBeFalsy();
+      await expectNeverToHappen(() => {
+        expect(
+          screen.getByText(
+            'Enter a finish time in the 24 hour clock format, for example, 08:00 or 0800'
+          )
+        ).toBeInTheDocument();
```
Comment (holly-blues) [inline src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx:432]: Why the change from waitFor and expect(errorMessage).toBeFalsy()?

Comment (LaurieBennett) [reply src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx:432]: This test wasn't actually working, me and Joanna tried a few things to get the test working and found this was the best way to test something not appearing on the screen.

**Comment by LaurieBennett** [comment]
> Just a question - does the backend handle these smart entries already?

Since our times are converted to dayJS dateTime the times are handled the same way as full 4 or 5 digit times, dayjs takes 2 digit times and converts them into appropriate times

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx (hunk 6 above)
Comment (holly-blues) [reply src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx:432]: Ok makes more sense, some sort of test library hell

File: src/components/common/validation/time-format/ValidatedTimeEntry.jsx (hunk 4 above)
Comment (LaurieBennett) [reply src/components/common/validation/time-format/ValidatedTimeEntry.jsx]: I've renamed this to a clearer name

File: src/components/common/validation/time-format/ValidatedTimeEntry.jsx [hunk 7]
```diff
@@ -8,7 +8,22 @@ const ValidatedTimeEntry = ({
...
+      return timeRegEx.test(time);
+    } else if (time.length > 3 && time.length < 6) {
+      const timeRegEx = /^([01]\d|2[0-3]):?([0-5]\d)$/;
+      return timeRegEx.test(time);
+    } else if (time.length == 0 && timeType === 'finish time') {
+      console.log('here');
```
[the remark this answers is not in the document]
Comment (LaurieBennett) [reply src/components/common/validation/time-format/ValidatedTimeEntry.jsx]: I've removed this console log

File: src/components/common/validation/time-format/ValidatedTimeEntry.jsx (hunk 5 above)
Comment (LaurieBennett) [reply src/components/common/validation/time-format/ValidatedTimeEntry.jsx]: I've renamed the regexes, there isn't currently a need to use these elsewhere as this is the only place we need to validate times.

**Review by holly-blues — APPROVED** [review]

**Review by daleelR — APPROVED** [review]
